### ISSUE TITLE
Data race in Batch.Reset()

### ISSUE THREAD

**Issue by kevgs** [blevesearch/bleve#260]
Here is a code to reproduce an issue.

i, _ := bleve.New("testidx", NewIndexMapping())
b := i.NewBatch()
b.Index("1", 1)
i.Batch(b)
b.Reset()

**Comment by mschoch** [comment]
Confirmed, output of race detector:

```
$ go test -v -run=TestBatchRace -race
=== RUN   TestBatchRace
==================
WARNING: DATA RACE
Write by goroutine 10:
  runtime.mapdelete()
      /Users/mschoch/Documents/research/gosrc/src/runtime/hashmap.go:511 +0x0
  github.com/blevesearch/bleve/index.(*Batch).Reset()
      /Users/mschoch/go/src/github.com/blevesearch/bleve/index/index.go:172 +0x10b
  github.com/blevesearch/bleve.(*Batch).Reset()
      /Users/mschoch/go/src/github.com/blevesearch/bleve/index.go:83 +0x45
  github.com/blevesearch/bleve.TestBatchRace()
      /Users/mschoch/go/src/github.com/blevesearch/bleve/index_test.go:1157 +0x253
  testing.tRunner()
      /Users/mschoch/Documents/research/gosrc/src/testing/testing.go:456 +0xdc

Previous read by goroutine 11:
  runtime.mapiternext()
      /Users/mschoch/Documents/research/gosrc/src/runtime/hashmap.go:621 +0x0
  github.com/blevesearch/bleve/index/upside_down.(*UpsideDownCouch).Batch.func1()
      /Users/mschoch/go/src/github.com/blevesearch/bleve/index/upside_down/upside_down.go:724 +0x2d8

Goroutine 10 (running) created at:
  testing.RunTests()
      /Users/mschoch/Documents/research/gosrc/src/testing/testing.go:561 +0xaa3
  testing.(*M).Run()
      /Users/mschoch/Documents/research/gosrc/src/testing/testing.go:494 +0xe4
  github.com/blevesearch/bleve.TestMain()
      /Users/mschoch/go/src/github.com/blevesearch/bleve/examples_test.go:30 +0xf9
  main.main()
      github.com/blevesearch/bleve/_test/_testmain.go:190 +0x209

Goroutine 11 (finished) created at:
  github.com/blevesearch/bleve/index/upside_down.(*UpsideDownCouch).Batch()
      /Users/mschoch/go/src/github.com/blevesearch/bleve/index/upside_down/upside_down.go:738 +0x25c
  github.com/blevesearch/bleve.(*indexImpl).Batch()
      /Users/mschoch/go/src/github.com/blevesearch/bleve/index_impl.go:310 +0x170
  github.com/blevesearch/bleve.TestBatchRace()
      /Users/mschoch/go/src/github.com/blevesearch/bleve/index_test.go:1156 +0x245
  testing.tRunner()
      /Users/mschoch/Documents/research/gosrc/src/testing/testing.go:456 +0xdc
==================
--- PASS: TestBatchRace (0.00s)
PASS
Found 1 data race(s)
exit status 66
FAIL	github.com/blevesearch/bleve	1.020s
```

**Comment by mschoch** [comment]
This issue was resolved as part of 	b8a2fbb
